### DISCUSSION TITLE
Raw queries documentation improvements

### DISCUSSION THREAD

**Discussion by janosimas** [rbatis/rbatis#294] [General]
Hi, great crate! I've been using for some experiments and it's quite simple!

The biggest issue I had was with documentation for raw queries, I had to try and error to be able to get just some fields of a table. I couldn't find an example in the docs that show the output as a `HashMap`:

```
#[sql(rb, "select name from foo where some_x = ?")]
pub async fn select_x(rb: &Rbatis, value: &str) -> HashMap<String, String> {}
```

**Comment by zhuxiujia** [comment]
sql and py_sql macro support hashmap。  use rb.fetch（）  also support hashmap。

if you use crud trait query table，must impl crud trait for hashmap  support json decode。

**Comment by zhuxiujia** [comment]
> Hi, great crate! I've been using for some experiments and it's quite simple!
> 
> The biggest issue I had was with documentation for raw queries, I had to try and error to be able to get just some fields of a table. I couldn't find an example in the docs that show the output as a `HashMap`:
> 
> ```
> #[sql(rb, "select name from foo where some_x = ?")]
> pub async fn select_x(rb: &Rbatis, value: &str) -> HashMap<String, String> {}
> ```

you can see  our  doc website   https://rbatis.github.io/rbatis.io/#/en/

**Comment by janosimas** [comment]
I've seen the documentation, what I meat is that it could be improved with an example a raw query for generic data.
The examples show raw queries for rich object types, like `BizActivity`, but what if I want to get only the names of `BizActivity` with some property?

**Comment by zhuxiujia** [comment]
> I've seen the documentation, what I meat is that it could be improved with an example a raw query for generic data.
> The examples show raw queries for rich object types, like `BizActivity`, but what if I want to get only the names of `BizActivity` with some property?

you can use  crud_table macro define only fetch some columns.  for example(query this struct will only fetch  id,name column):
```rust  
    #[crud_table(table_columns:"id,name")]
    #[derive(Clone, Debug)]
    pub struct BizActivity {
      pub id:String,
      pub name:Option<String>,
        pub age:Option<i32>,
}
```

**Comment by janosimas** [comment]
That's a cool feature I didn't notice.

About my question, I may not have been clear.
I my case, I have a `BizActivity` in my database but it's a heavy structure. I don't want to load it unless I need.
For example, if I want to show to the user a list of the available `BizActivity`, I just want to query for the names. After the user selects one of them, I'll load the full data for `BizActivity`.

This is not a complaint, I was able to do this. But the documentation was not so clear for this kind of queries that don't map to a full object.

**Comment by zhuxiujia** [comment]
> That's a cool feature I didn't notice.
> 
> About my question, I may not have been clear.
> I my case, I have a `BizActivity` in my database but it's a heavy structure. I don't want to load it unless I need.
> For example, if I want to show to the user a list of the available `BizActivity`, I just want to query for the names. After the user selects one of them, I'll load the full data for `BizActivity`.
> 
> This is not a complaint, I was able to do this. But the documentation was not so clear for this kind of queries that don't map to a full object.

My understanding is that the structure of the database layer and the structure of the presentation layer must conform to a certain paradigm. I used to be a Java developer, and for this kind of situation we usually use DDD (Domain Driven Design)
```rust
    #[crud_table(table_columns:"id,name")]
    #[derive(Clone, Debug)]
    pub struct BizActivity {
      pub id:String,
      pub name:Option<String>
    }


pub struct BizActivityVO {
      pub name:Option<String>
 }

impl From<BizActivity > for BizActivityVO {
    
} 


```



or  

you can use this define:
```rust
    #[crud_table(table_columns:"name")]
    #[derive(Clone, Debug)]
    pub struct BizActivityOnlyName {
      pub name:Option<String>
    }
```

**Comment by janosimas** [comment]
It feels like an overkill for me but I get the idea. thanks.
